app/models/amazon_agent_manager.rb

```diff
@@ -0,0 +1,411 @@
+        ]
+      ).security_groups
+      return sgs[0].group_id if sgs.length > 0
+
+      # create security group if not exist
+      security_group = @ec2.create_security_group({
+        group_name: group_name,
+        description: 'Security group for MIQ SSA Agent',
+        vpc_id: @ec2.client.describe_vpcs.vpcs[0].vpc_id
+      })
+
+      security_group.authorize_ingress({
+        ip_permissions: [{
+          ip_protocol: 'tcp',
+          from_port: 22,
+          to_port: 22,
+          ip_ranges: [{
+            cidr_ip: '0.0.0.0/0'
+          }]}]
+      })
+
+      security_group.authorize_ingress({
+        ip_permissions: [{
+          ip_protocol: 'tcp',
+          from_port: 80,
+          to_port: 80,
+          ip_ranges: [{
+            cidr_ip: '0.0.0.0/0'
+          }]}]
+      })
+
+      security_group.authorize_ingress({
+        ip_permissions: [{
+          ip_protocol: 'tcp',
+          from_port: 443,
+          to_port: 443,
+          ip_ranges: [{
+            cidr_ip: '0.0.0.0/0'
+          }]}]
+      })
+
+      security_group.group_id
+    end
+  end
+
+  # Get Key Pair for SSH. Create a new one if not exists.
+  def get_key_pair(pair_name = MIQ_SSA)
+    kp = Authentication.where(name: pair_name)
+    return kp[0] if kp.length > 0 && kp[0].resource_id == ems.id
+
+    ManageIQ::Providers::CloudManager::AuthKeyPair.create_key_pair(ems.id,
+      { :key_name => pair_name })
+  end
+
+  def create_pem_file(pair_name = MIQ_SSA)
+    kp = get_key_pair(pair_name)
+    pem_file_name = pair_name+".pem"
+    File.open(pem_file_name, 'w') {|f| f.write(kp.auth_key) }
+    File.chmod(0400, pem_file_name)
+    pem_file_name
+  end
+
+  def setup_ruby(log = '/var/log/miq_ssa_deploy.log')
+    <<~SETUP
+      yum -y update > #{log} 2>&1
+      yum -y install git-core zlib zlib-devel gcc-c++ patch readline readline-devel libyaml-devel libffi-devel openssl-devel make bzip2 autoconf automake libtool bison curl sqlite-devel postgresql-devel >> #{log} 2>&1
+      git clone https://github.com/rbenv/rbenv.git ~/.rbenv >> #{log} 2>&1
+      export HOME="/root" >> #{log} 2>&1
+      echo 'export HOME="/root"' >> ~/.bash_profile
+      echo 'export PATH="$HOME/.rbenv/bin:$PATH"' >> ~/.bash_profile
+      echo 'eval "$(/root/.rbenv/bin/rbenv init -)"' >> ~/.bash_profile
+      source ~/.bash_profile >> #{log} 2>&1
+      git clone git://github.com/rbenv/ruby-build.git ~/.rbenv/plugins/ruby-build >> #{log} 2>&1
+      echo 'export PATH="$HOME/.rbenv/plugins/ruby-build/bin:$PATH"' >> ~/.bash_profile
+      source ~/.bash_profile >> #{log} 2>&1
+      echo $PATH >> #{log} 2>&1
+      rbenv install -l >> #{log} 2>&1
+      rbenv install 2.3.3 >> #{log} 2>&1
+      rbenv global 2.3.3 >> #{log} 2>&1
+      gem install bundler >> #{log} 2>&1
+      gem install rails >> #{log} 2>&1
+      gem install aws-sdk >> #{log} 2>&1
+      rbenv rehash >> #{log} 2>&1
+      ruby -v >> #{log} 2>&1
+    SETUP
+  end
+
+  def default_settings
```

Comment:
I think generating yaml files this way is fragile. The shell script should probably call a ruby script that does a `YAML.dump` on an initialized `Hash`.